### PR TITLE
Backport: Removal of regex usage on denom validation 

### PR DESCRIPTION
**Pull request by mattverse**
## Description

Closes: #XXXX

Backports removal of regex usage on denom validation upstream
---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

## Summary by CodeRabbit

- **Refactor**
	- Improved handling and validation of coin denominations.
- **Tests**
	- Enhanced testing for coin denomination validation.
- **New Features**
	- Added custom parsing logic for coin denominations.
	- Introduced a function to parse amount and denomination from a coin string.

### CODE REVIEW

File: types/coin.go [hunk 1]
```diff
@@ -871,12 +857,49 @@ func SetCoinDenomRegex(reFn func() string) {
...
 		return fmt.Errorf("invalid denom: %s", denom)
 	}
+
 	return nil
 }
 
+// isValidRune checks if a given rune is a valid character for a rune.
+// It returns true if the rune is a letter, digit, '/', ':', '.', '_', or '-'.
+func isValidRune(r rune) bool {
+	return unicode.IsLetter(r) || unicode.IsDigit(r) || r == '/' || r == ':' || r == '.' || r == '_' || r == '-'
```
Comment (ValarDragon) [inline types/coin.go:877]: btw we should change this upstream to just be bytes, and a couple of if statements. Rune's / these unicode calls are even more complex than what we need. 

e.g. imagine checking `[0-9/_]` as `byte >= 0x2F && byte <= 0x3A`.

Comment (ValarDragon) [reply types/coin.go:877]: Can make an issue for trying this upstream

**Review by ValarDragon — APPROVED** [review]
Awesome!

**Comment by czarcas7ic** [comment]
Setting as draft until tests are fixed

**Comment by mattverse** [comment]
linting is failing but tbh idk why + im not sure if we want to create divergance with upstream just for this lint 🤔 
@czarcas7ic thoughts?

**Comment by ValarDragon** [comment]
I agree, idc about linting

**Comment by czarcas7ic** [comment]
I mean, can we add an ignore line so every future PR doesnt fail

**Comment by czarcas7ic** [comment]
Its just saying to use switch instead of if else, should be able to add a gocritic ignore

**Comment by mattverse** [comment]
@czarcas7ic ok, I actually can't fix the lint (`nolint` is not working for some reason, also cannot replicate this lint error in my local env), I would rather fix our lint CI to something upstream sdk has, thoughts?

**Comment by czarcas7ic** [comment]
Yeah I am fine with any solution that keeps a lint CI check and passed

**Pull request merged**